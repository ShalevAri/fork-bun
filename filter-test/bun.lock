{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "root",
      "dependencies": {
        "lodash": "^4.17.21",
      },
    },
    "packages/pkg1": {
      "name": "pkg1",
      "version": "0.0.1",
    },
    "packages/pkg2": {
      "name": "pkg2",
      "version": "0.0.1",
    },
  },
  "packages": {
    "lodash": ["lodash@4.17.21", "", {}, "sha512-v2kDEe57lecTulaDIuNTPy3Ry4gLGJ6Z1O3vE1krgXZNrsQ+LFTGHVxVjcXPs17LhbZVGedAJv8XZ1tvj5FvSg=="],

    "pkg1": ["pkg1@workspace:packages/pkg1"],

    "pkg2": ["pkg2@workspace:packages/pkg2"],
  }
}
