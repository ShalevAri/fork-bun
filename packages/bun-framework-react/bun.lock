{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "dependencies": {
        "react": "19.2.0-canary-67a44bcd-20250915",
        "react-dom": "19.2.0-canary-67a44bcd-20250915",
        "react-refresh": "^0.17.0",
        "react-server-dom-bun": "^0.0.0-experimental-603e6108-20241029",
      },
      "devDependencies": {
        "@types/react": "^19.1.13",
        "@types/react-dom": "^19.1.9",
      },
    },
  },
  "packages": {
    "@types/react": ["@types/react@19.1.13", "", { "dependencies": { "csstype": "^3.0.2" } }, "sha512-hHkbU/eoO3EG5/MZkuFSKmYqPbSVk5byPFa3e7y/8TybHiLMACgI8seVYlicwk7H5K/rI2px9xrQp/C+AUDTiQ=="],

    "@types/react-dom": ["@types/react-dom@19.1.9", "", { "peerDependencies": { "@types/react": "^19.0.0" } }, "sha512-qXRuZaOsAdXKFyOhRBg6Lqqc0yay13vN7KrIg4L7N4aaHN68ma9OK3NE1BoDFgFOTfM7zg+3/8+2n8rLUH3OKQ=="],

    "csstype": ["csstype@3.1.3", "", {}, "sha512-M1uQkMl8rQK/szD0LNhtqxIPLpimGm8sOBwU7lLnCpSbTyY3yeU1Vc7l4KT5zT4s/yOxHH5O7tIuuLOCnLADRw=="],

    "neo-async": ["neo-async@2.6.2", "", {}, "sha512-Yd3UES5mWCSqR+qNT93S3UoYUkqAZ9lLg8a7g9rimsWmYGK8cVToA4/sF3RrshdyV3sAGMXVUmpMYOw+dLpOuw=="],

    "react": ["react@19.2.0-canary-67a44bcd-20250915", "", {}, "sha512-93GXCSAJhSkIpC0Z8adssVHP/MIgmufZ0eFBBP3TUWuLYNFk7Pv8Sal6UZ/UGVeX+KZPSu95vSR5guUvqtGrDQ=="],

    "react-dom": ["react-dom@19.2.0-canary-67a44bcd-20250915", "", { "dependencies": { "scheduler": "0.27.0-canary-67a44bcd-20250915" }, "peerDependencies": { "react": "19.2.0-canary-67a44bcd-20250915" } }, "sha512-s+aP2nMr/wKdIrf0Z/fyYObl8IyMXbaW+ujekxG7hKtssuWjOJ2OhS9bg34/DFm1OaMczKKyW2Crv0BYRUlJfA=="],

    "react-refresh": ["react-refresh@0.17.0", "", {}, "sha512-z6F7K9bV85EfseRCp2bzrpyQ0Gkw1uLoCel9XBVWPg/TjRj94SkJzUTGfOa4bs7iJvBWtQG0Wq7wnI0syw3EBQ=="],

    "react-server-dom-bun": ["react-server-dom-bun@0.0.0-experimental-603e6108-20241029", "", { "dependencies": { "neo-async": "^2.6.1" } }, "sha512-FfteCHlOgJSnDJRatgIkIU74jQQ9M1+fH2e6kfY9Sibu8FAWEUjgApKQPDfiXgjrkY7w0ITQu0b2FezC0eGzCw=="],

    "scheduler": ["scheduler@0.27.0-canary-67a44bcd-20250915", "", {}, "sha512-smZ2mJWYQBKEpnEHOavM3YwhduC6pD4esZTefLPwQok/Ca65W/tWDHRHJhFHYi+A8ottJJA8G6tXfhA2wuGH0A=="],
  }
}
