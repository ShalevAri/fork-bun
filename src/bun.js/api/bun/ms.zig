const std = @import("std");
const bun = @import("bun");
const jsc = bun.jsc;
const JSValue = jsc.JSValue;
const JSGlobalObject = jsc.JSGlobalObject;

const MS_PER_SECOND = 1000.0;
const MS_PER_MINUTE = MS_PER_SECOND * 60.0;
const MS_PER_HOUR = MS_PER_MINUTE * 60.0;
const MS_PER_DAY = MS_PER_HOUR * 24.0;
const MS_PER_WEEK = MS_PER_DAY * 7.0;
const MS_PER_YEAR = MS_PER_DAY * 365.25;

/// Parse a time string like "2d", "1.5h", "5m" to milliseconds
pub fn parse(input: []const u8) ?f64 {
    if (input.len == 0 or input.len > 100) return null;

    var i: usize = 0;
    var negative = false;

    // Skip leading whitespace
    while (i < input.len and std.ascii.isWhitespace(input[i])) : (i += 1) {}
    if (i >= input.len) return null;

    // Check for negative sign
    if (input[i] == '-') {
        negative = true;
        i += 1;
    }

    // Parse number
    const start = i;
    var has_digits = false;
    var has_dot = false;

    while (i < input.len) {
        const c = input[i];
        if (std.ascii.isDigit(c)) {
            has_digits = true;
            i += 1;
        } else if (c == '.' and !has_dot) {
            has_dot = true;
            i += 1;
        } else if (std.ascii.isWhitespace(c) or std.ascii.isAlphabetic(c)) {
            break;
        } else {
            return null;
        }
    }

    if (!has_digits) return null;

    const number_str = input[start..i];
    var value = std.fmt.parseFloat(f64, number_str) catch return null;

    if (negative) value = -value;

    // Skip whitespace after number
    while (i < input.len and std.ascii.isWhitespace(input[i])) : (i += 1) {}

    // Get unit (rest of string, excluding trailing whitespace)
    var unit_end = input.len;
    while (unit_end > i and std.ascii.isWhitespace(input[unit_end - 1])) : (unit_end -= 1) {}
    const unit = input[i..unit_end];

    // Default to milliseconds if no unit
    if (unit.len == 0) return value;

    // Match unit (case-insensitive)
    return if (getMultiplier(unit)) |m| value * m else null;
}

fn getMultiplier(unit: []const u8) ?f64 {
    // Years
    if (std.ascii.eqlIgnoreCase(unit, "years") or std.ascii.eqlIgnoreCase(unit, "year") or
        std.ascii.eqlIgnoreCase(unit, "yrs") or std.ascii.eqlIgnoreCase(unit, "yr") or
        std.ascii.eqlIgnoreCase(unit, "y"))
    {
        return MS_PER_YEAR;
    }

    // Weeks
    if (std.ascii.eqlIgnoreCase(unit, "weeks") or std.ascii.eqlIgnoreCase(unit, "week") or
        std.ascii.eqlIgnoreCase(unit, "w"))
    {
        return MS_PER_WEEK;
    }

    // Days
    if (std.ascii.eqlIgnoreCase(unit, "days") or std.ascii.eqlIgnoreCase(unit, "day") or
        std.ascii.eqlIgnoreCase(unit, "d"))
    {
        return MS_PER_DAY;
    }

    // Hours
    if (std.ascii.eqlIgnoreCase(unit, "hours") or std.ascii.eqlIgnoreCase(unit, "hour") or
        std.ascii.eqlIgnoreCase(unit, "hrs") or std.ascii.eqlIgnoreCase(unit, "hr") or
        std.ascii.eqlIgnoreCase(unit, "h"))
    {
        return MS_PER_HOUR;
    }

    // Minutes
    if (std.ascii.eqlIgnoreCase(unit, "minutes") or std.ascii.eqlIgnoreCase(unit, "minute") or
        std.ascii.eqlIgnoreCase(unit, "mins") or std.ascii.eqlIgnoreCase(unit, "min") or
        std.ascii.eqlIgnoreCase(unit, "m"))
    {
        return MS_PER_MINUTE;
    }

    // Seconds
    if (std.ascii.eqlIgnoreCase(unit, "seconds") or std.ascii.eqlIgnoreCase(unit, "second") or
        std.ascii.eqlIgnoreCase(unit, "secs") or std.ascii.eqlIgnoreCase(unit, "sec") or
        std.ascii.eqlIgnoreCase(unit, "s"))
    {
        return MS_PER_SECOND;
    }

    // Milliseconds
    if (std.ascii.eqlIgnoreCase(unit, "milliseconds") or std.ascii.eqlIgnoreCase(unit, "millisecond") or
        std.ascii.eqlIgnoreCase(unit, "msecs") or std.ascii.eqlIgnoreCase(unit, "msec") or
        std.ascii.eqlIgnoreCase(unit, "ms"))
    {
        return 1.0;
    }

    return null;
}


/// Format milliseconds to a human-readable string
pub fn format(allocator: std.mem.Allocator, ms: f64, long: bool) ![]const u8 {
    const abs_ms = @abs(ms);

    if (abs_ms >= MS_PER_DAY) {
        const days = @round(ms / MS_PER_DAY);
        const days_int = @as(i64, @intFromFloat(days));
        if (long) {
            const plural = abs_ms >= MS_PER_DAY * 1.5;
            return std.fmt.allocPrint(allocator, "{d} day{s}", .{ days_int, if (plural) "s" else "" });
        }
        return std.fmt.allocPrint(allocator, "{d}d", .{days_int});
    }

    if (abs_ms >= MS_PER_HOUR) {
        const hours = @round(ms / MS_PER_HOUR);
        const hours_int = @as(i64, @intFromFloat(hours));
        if (long) {
            const plural = abs_ms >= MS_PER_HOUR * 1.5;
            return std.fmt.allocPrint(allocator, "{d} hour{s}", .{ hours_int, if (plural) "s" else "" });
        }
        return std.fmt.allocPrint(allocator, "{d}h", .{hours_int});
    }

    if (abs_ms >= MS_PER_MINUTE) {
        const minutes = @round(ms / MS_PER_MINUTE);
        const minutes_int = @as(i64, @intFromFloat(minutes));
        if (long) {
            const plural = abs_ms >= MS_PER_MINUTE * 1.5;
            return std.fmt.allocPrint(allocator, "{d} minute{s}", .{ minutes_int, if (plural) "s" else "" });
        }
        return std.fmt.allocPrint(allocator, "{d}m", .{minutes_int});
    }

    if (abs_ms >= MS_PER_SECOND) {
        const seconds = @round(ms / MS_PER_SECOND);
        const seconds_int = @as(i64, @intFromFloat(seconds));
        if (long) {
            const plural = abs_ms >= MS_PER_SECOND * 1.5;
            return std.fmt.allocPrint(allocator, "{d} second{s}", .{ seconds_int, if (plural) "s" else "" });
        }
        return std.fmt.allocPrint(allocator, "{d}s", .{seconds_int});
    }

    const ms_int = @as(i64, @intFromFloat(ms));
    if (long) {
        return std.fmt.allocPrint(allocator, "{d} ms", .{ms_int});
    }
    return std.fmt.allocPrint(allocator, "{d}ms", .{ms_int});
}

/// JavaScript function: Bun.ms(value, options?)
pub fn jsFunction(
    globalThis: *JSGlobalObject,
    callframe: *jsc.CallFrame,
) bun.JSError!jsc.JSValue {
    const args = callframe.arguments_old(2);
    if (args.len == 0) return .js_undefined;

    const input = args.ptr[0];

    // If input is a number, format it to a string
    if (input.isNumber()) {
        const ms_value = input.asNumber();

        if (std.math.isNan(ms_value) or std.math.isInf(ms_value)) {
            return globalThis.throwInvalidArguments("Value must be a finite number", .{});
        }

        // Check for options argument
        var long = false;
        if (args.len > 1 and args.ptr[1].isObject()) {
            const options = args.ptr[1];
            if (try options.get(globalThis, "long")) |long_value| {
                long = long_value.toBoolean();
            }
        }

        const result = format(bun.default_allocator, ms_value, long) catch {
            return globalThis.throwOutOfMemory();
        };
        defer bun.default_allocator.free(result);

        return bun.String.fromBytes(result).toJS(globalThis);
    }

    // If input is a string, parse it to milliseconds
    if (input.isString()) {
        const str = try input.getZigString(globalThis);
        const slice = str.toSlice(bun.default_allocator);
        defer slice.deinit();

        const result = parse(slice.slice()) orelse return .js_undefined;
        return JSValue.jsNumber(result);
    }

    // Invalid input type
    return .js_undefined;
}
