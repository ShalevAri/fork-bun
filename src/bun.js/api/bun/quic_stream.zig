const std = @import("std");
const bun = @import("../../../bun.zig");
const jsc = bun.jsc;
const uws = bun.uws;
const Environment = bun.Environment;
const Async = bun.Async;

const log = bun.Output.scoped(.QuicStream, .hidden);

pub const QuicStream = struct {
    const This = @This();

    // JavaScript class bindings
    pub const js = jsc.Codegen.JSQuicStream;
    pub const toJS = js.toJS;
    pub const fromJS = js.fromJS;
    pub const fromJSDirect = js.fromJSDirect;

    pub const new = bun.TrivialNew(@This());

    const RefCount = bun.ptr.RefCount(@This(), "ref_count", deinit, .{});
    pub const ref = RefCount.ref;
    pub const deref = RefCount.deref;

    // The underlying lsquic stream
    stream: ?*uws.quic.Stream = null,
    
    // Reference to parent socket
    socket: *QuicSocket,
    
    // Stream ID
    stream_id: u64,
    
    // Optional data attached to the stream
    data_value: jsc.JSValue = .zero,
    
    // JavaScript this value
    this_value: jsc.JSValue = .zero,
    
    // Reference counting
    ref_count: RefCount,
    poll_ref: Async.KeepAlive = Async.KeepAlive.init(),
    
    // Stream state
    flags: Flags = .{},
    
    // Buffered writes before stream is connected
    write_buffer: std.ArrayList([]const u8) = undefined,
    write_buffer_initialized: bool = false,
    write_buffer_mutex: std.Thread.Mutex = .{},
    
    has_pending_activity: std.atomic.Value(bool) = std.atomic.Value(bool).init(true),

    pub const Flags = packed struct {
        is_readable: bool = true,
        is_writable: bool = true,
        is_closed: bool = false,
        has_backpressure: bool = false,
        fin_sent: bool = false,
        fin_received: bool = false,
        _: u26 = 0,
    };

    pub fn hasPendingActivity(this: *This) callconv(.C) bool {
        return this.has_pending_activity.load(.acquire);
    }

    pub fn memoryCost(_: *This) usize {
        return @sizeOf(This);
    }

    pub fn finalize(this: *This) void {
        this.deinit();
    }

    pub fn deinit(this: *This) void {
        this.poll_ref.unref(jsc.VirtualMachine.get());

        // Clean up write buffer
        if (this.write_buffer_initialized) {
            this.write_buffer_mutex.lock();
            defer this.write_buffer_mutex.unlock();
            
            // Free any buffered write data
            for (this.write_buffer.items) |buffered_data| {
                bun.default_allocator.free(buffered_data);
            }
            this.write_buffer.deinit();
            this.write_buffer_initialized = false;
        }

        // Unprotect the data value if set
        if (!this.data_value.isEmptyOrUndefinedOrNull()) {
            this.data_value.unprotect();
            this.data_value = .zero;
        }

        // Close stream if still open
        if (this.stream != null and !this.flags.is_closed) {
            this.closeImpl();
        }

        // Deref the parent socket
        this.socket.deref();
    }

    // Initialize a new QUIC stream
    pub fn init(allocator: std.mem.Allocator, socket: *QuicSocket, stream_id: u64, data_value: jsc.JSValue) !*This {
        const this = try allocator.create(This);
        this.* = This{
            .ref_count = RefCount.init(),
            .socket = socket,
            .stream_id = stream_id,
            .data_value = data_value,
        };
        
        // Initialize write buffer
        this.write_buffer = std.ArrayList([]const u8).init(allocator);
        this.write_buffer_initialized = true;
        
        // Ref the parent socket to keep it alive
        socket.ref();
        
        // Protect the data value if set
        if (!data_value.isEmptyOrUndefinedOrNull()) {
            data_value.protect();
        }
        
        this.ref();
        return this;
    }

    // Write data to the stream
    pub fn write(this: *This, globalObject: *jsc.JSGlobalObject, callframe: *jsc.CallFrame) bun.JSError!jsc.JSValue {
        const arguments = callframe.arguments_old(1);
        if (arguments.len < 1) {
            return globalObject.throwInvalidArguments("write() requires a buffer argument", .{});
        }

        if (this.flags.is_closed) {
            return globalObject.throwInvalidArguments("Stream is closed", .{});
        }

        const data = arguments.ptr[0];
        
        // Convert to buffer
        var buffer: []const u8 = undefined;
        if (data.asArrayBuffer(globalObject)) |array_buffer| {
            buffer = array_buffer.slice();
        } else if (data.isString()) {
            const str = try data.toBunString(globalObject);
            defer str.deref();
            const utf8 = str.toUTF8(bun.default_allocator);
            defer utf8.deinit();
            buffer = utf8.slice();
        } else {
            return globalObject.throwInvalidArguments("write() expects a Buffer or string", .{});
        }

        return this.writeInternal(buffer, globalObject);
    }
    
    // Internal write method that can be called from both JS and internal code
    fn writeInternal(this: *This, buffer: []const u8, globalObject: ?*jsc.JSGlobalObject) bun.JSError!jsc.JSValue {
        // Write to the underlying stream or buffer if stream not yet connected
        if (this.stream) |stream| {
            log("QuicStream.write: Writing {} bytes directly to connected stream {*} (ID: {})", .{buffer.len, stream, this.stream_id});
            const written = stream.write(buffer);
            const written_usize: usize = if (written >= 0) @intCast(written) else 0;
            log("QuicStream.write: stream.write returned {} bytes for stream {}", .{written, this.stream_id});
            
            // Handle backpressure - if not all data was written, set backpressure flag
            if (written_usize < buffer.len) {
                this.flags.has_backpressure = true;
                log("QuicStream.write: backpressure detected on stream {}, wrote {} of {} bytes", .{ this.stream_id, written_usize, buffer.len });
            } else {
                this.flags.has_backpressure = false;
            }
            
            log("QuicStream.write: wrote {} bytes to stream {}", .{ written_usize, this.stream_id });
            const written_float: f64 = @floatFromInt(written_usize);
            return jsc.JSValue.jsNumber(written_float);
        } else {
            // Stream not connected yet, buffer the write
            log("QuicStream.write: Stream {} not connected, attempting to buffer {} bytes", .{this.stream_id, buffer.len});
            if (!this.write_buffer_initialized) {
                log("QuicStream.write: write buffer not initialized for stream {}, returning 0", .{this.stream_id});
                return jsc.JSValue.jsNumber(0);
            }
            
            this.write_buffer_mutex.lock();
            defer this.write_buffer_mutex.unlock();
            
            // Make a copy of the data to buffer
            const buffered_data = bun.default_allocator.dupe(u8, buffer) catch |err| {
                log("QuicStream.write: failed to allocate buffer memory for stream {}: {}", .{this.stream_id, err});
                if (globalObject) |globalObj| {
                    return globalObj.throwError(err, "Failed to allocate memory for write buffer");
                } else {
                    return jsc.JSValue.jsNumber(0);
                }
            };
            
            // Add to write buffer
            this.write_buffer.append(buffered_data) catch |err| {
                bun.default_allocator.free(buffered_data);
                log("QuicStream.write: failed to append to write buffer for stream {}: {}", .{this.stream_id, err});
                if (globalObject) |globalObj| {
                    return globalObj.throwError(err, "Failed to buffer write data");
                } else {
                    return jsc.JSValue.jsNumber(0);
                }
            };
            
            log("QuicStream.write: buffered {} bytes for stream {} (buffer size: {})", .{ buffer.len, this.stream_id, this.write_buffer.items.len });
            
            // Return the buffered size so caller thinks the write succeeded
            const buffered_float: f64 = @floatFromInt(buffer.len);
            return jsc.JSValue.jsNumber(buffered_float);
        }
    }

    // Buffer write data when stream is not yet connected (internal method)
    pub fn bufferWrite(this: *This, data: []const u8) !void {
        if (this.flags.is_closed) return error.StreamClosed;
        
        if (!this.write_buffer_initialized) {
            return error.BufferNotInitialized;
        }
        
        this.write_buffer_mutex.lock();
        defer this.write_buffer_mutex.unlock();
        
        // Make a copy of the data to buffer
        const buffered_data = try bun.default_allocator.dupe(u8, data);
        errdefer bun.default_allocator.free(buffered_data);
        
        // Add to write buffer
        try this.write_buffer.append(buffered_data);
        
        log("bufferWrite: buffered {} bytes for stream {} (buffer size: {})", .{ data.len, this.stream_id, this.write_buffer.items.len });
    }

    // End the stream (graceful close with FIN)
    pub fn end(this: *This, _: *jsc.JSGlobalObject, _: *jsc.CallFrame) bun.JSError!jsc.JSValue {
        if (this.flags.is_closed or this.flags.fin_sent) {
            return .js_undefined;
        }

        if (this.stream) |stream| {
            this.flags.fin_sent = true;
            _ = stream.shutdown(); // Shutdown write side
            log("QuicStream.end: sent FIN on stream {}", .{this.stream_id});
        }

        return .js_undefined;
    }

    // Close the stream immediately
    pub fn close(this: *This, _: *jsc.JSGlobalObject, _: *jsc.CallFrame) bun.JSError!jsc.JSValue {
        this.closeImpl();
        return .js_undefined;
    }

    fn closeImpl(this: *This) void {
        if (this.flags.is_closed) return;
        
        this.flags.is_closed = true;
        this.has_pending_activity.store(false, .release);

        if (this.stream) |stream| {
            // Remove from socket's stream mapping before closing
            _ = this.socket.removeStreamMapping(stream);
            
            stream.close();
            this.stream = null;
            log("QuicStream.close: closed stream {}", .{this.stream_id});
        }

        // Clear any remaining buffered writes
        if (this.write_buffer_initialized) {
            this.write_buffer_mutex.lock();
            defer this.write_buffer_mutex.unlock();
            
            for (this.write_buffer.items) |buffered_data| {
                bun.default_allocator.free(buffered_data);
            }
            this.write_buffer.clearAndFree();
        }
    }

    // Flush any buffered writes to the now-connected stream
    pub fn flushBufferedWrites(this: *This) void {
        log("flushBufferedWrites: stream_id={}, stream={*}, initialized={}, buffer_len={}", .{
            this.stream_id,
            this.stream, 
            this.write_buffer_initialized,
            if (this.write_buffer_initialized) this.write_buffer.items.len else 0
        });
        
        if (!this.write_buffer_initialized or this.stream == null) {
            log("flushBufferedWrites: early return for stream {} - not initialized or no stream", .{this.stream_id});
            return;
        }

        this.write_buffer_mutex.lock();
        defer this.write_buffer_mutex.unlock();

        const stream = this.stream.?;
        var total_written: usize = 0;
        var failed_writes: usize = 0;

        const buffer_count = this.write_buffer.items.len;
        log("flushBufferedWrites: flushing {} buffered writes to stream {*} (ID: {})", .{buffer_count, stream, this.stream_id});

        // Write all buffered data to the stream
        for (this.write_buffer.items) |buffered_data| {
            const written = stream.write(buffered_data);
            const written_usize: usize = if (written >= 0) @intCast(written) else 0;
            total_written += written_usize;
            
            if (written_usize < buffered_data.len) {
                this.flags.has_backpressure = true;
                failed_writes += 1;
                log("QuicStream.flushBufferedWrites: partial write {} of {} bytes for stream {}", .{ written_usize, buffered_data.len, this.stream_id });
            } else {
                log("QuicStream.flushBufferedWrites: wrote {} bytes for stream {}", .{ written_usize, this.stream_id });
            }
        }

        // Free the buffered data and clear the buffer
        for (this.write_buffer.items) |buffered_data| {
            bun.default_allocator.free(buffered_data);
        }
        this.write_buffer.clearRetainingCapacity();

        if (failed_writes > 0) {
            log("QuicStream.flushBufferedWrites: {} of {} buffered writes had backpressure for stream {}", .{ failed_writes, buffer_count, this.stream_id });
        } else {
            log("QuicStream.flushBufferedWrites: flushed {} buffered writes ({} total bytes) for stream {}", .{ buffer_count, total_written, this.stream_id });
        }
    }

    // JavaScript ref/unref for keeping the event loop alive
    pub fn jsRef(this: *This, _: *jsc.JSGlobalObject, _: *jsc.CallFrame) bun.JSError!jsc.JSValue {
        this.ref();
        this.poll_ref.ref(jsc.VirtualMachine.get());
        return .js_undefined;
    }

    pub fn jsUnref(this: *This, _: *jsc.JSGlobalObject, _: *jsc.CallFrame) bun.JSError!jsc.JSValue {
        this.poll_ref.unref(jsc.VirtualMachine.get());
        this.deref();
        return .js_undefined;
    }

    // Getters for JavaScript properties
    pub fn getId(this: *This, _: *jsc.JSGlobalObject) jsc.JSValue {
        const id_float: f64 = @floatFromInt(this.stream_id);
        return jsc.JSValue.jsNumber(id_float);
    }

    pub fn getSocket(this: *This, globalObject: *jsc.JSGlobalObject) jsc.JSValue {
        return this.socket.toJS(globalObject);
    }

    pub fn getData(this: *This, _: *jsc.JSGlobalObject) jsc.JSValue {
        return this.data_value;
    }

    pub fn getReadyState(this: *This, _: *jsc.JSGlobalObject) jsc.JSValue {
        if (this.flags.is_closed) {
            return jsc.JSValue.jsNumberFromChar(3); // CLOSED
        } else if (this.flags.fin_sent) {
            return jsc.JSValue.jsNumberFromChar(2); // CLOSING
        } else {
            return jsc.JSValue.jsNumberFromChar(1); // OPEN
        }
    }
};

// Import QuicSocket type
const QuicSocket = @import("quic_socket.zig").QuicSocket;