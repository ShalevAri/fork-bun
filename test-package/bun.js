export const value = "from-bun";
export const message = "This is from the Bun-specific export!";