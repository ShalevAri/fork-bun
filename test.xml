<root><message>Hello XML\!</message></root>
