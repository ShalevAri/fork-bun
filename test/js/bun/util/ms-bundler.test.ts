import { test, expect } from "bun:test";
import { bunExe, tempDir } from "harness";

// Test that Bun.ms works correctly in bundled code
test("Bun.ms works in bundled code", async () => {
  using dir = tempDir("ms-bundler-test", {
    "entry.ts": `
      // Test various Bun.ms calls
      console.log(JSON.stringify({
        oneSecond: Bun.ms("1s"),
        oneMinute: Bun.ms("1m"),
        oneHour: Bun.ms("1h"),
        oneDay: Bun.ms("1d"),
        twoWeeks: Bun.ms("2w"),
        halfYear: Bun.ms("0.5y"),
        withSpaces: Bun.ms("5 minutes"),
        negative: Bun.ms("-10s"),
        decimal: Bun.ms("1.5h"),
        justNumber: Bun.ms("100"),
        caseInsensitive: Bun.ms("2D"),
        formatShort: Bun.ms(1000),
        formatLong: Bun.ms(60000, { long: true }),
      }));
    `,
  });

  // Bundle the code
  await using buildProc = Bun.spawn({
    cmd: [bunExe(), "build", "entry.ts", "--outdir", ".", "--minify"],
    cwd: String(dir),
    stderr: "pipe",
    stdout: "pipe",
  });

  await buildProc.exited;
  expect(buildProc.exitCode).toBe(0);

  // Run the bundled code
  await using runProc = Bun.spawn({
    cmd: [bunExe(), "entry.js"],
    cwd: String(dir),
    stderr: "pipe",
    stdout: "pipe",
  });

  const [stdout, stderr] = await Promise.all([
    runProc.stdout.text(),
    runProc.stderr.text(),
    runProc.exited,
  ]);

  expect(runProc.exitCode).toBe(0);
  expect(stderr).toBe("");

  const result = JSON.parse(stdout.trim());

  // Verify all values are correct
  expect(result).toEqual({
    oneSecond: 1000,
    oneMinute: 60000,
    oneHour: 3600000,
    oneDay: 86400000,
    twoWeeks: 1209600000,
    halfYear: 15778800000,
    withSpaces: 300000,
    negative: -10000,
    decimal: 5400000,
    justNumber: 100,
    caseInsensitive: 172800000,
    formatShort: "1s",
    formatLong: "1 minute",
  });
});

// Note: Compile-time inlining of Bun.ms("literal") would be a future optimization
// For now, this test verifies that Bun.ms works correctly at runtime in bundled code
