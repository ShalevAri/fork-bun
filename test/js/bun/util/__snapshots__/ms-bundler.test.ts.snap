// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`bundler inlines Bun.ms with string literals 1`] = `
"var q=Bun.ms("1s"),x=Bun.ms("1m"),z=Bun.ms("1h"),A=Bun.ms("1d"),B=Bun.ms("2w"),C=Bun.ms("0.5y"),E=Bun.ms("5 minutes"),F=Bun.ms("-10s"),G=Bun.ms("1.5h"),J=Bun.ms("100"),K=Bun.ms("2D"),L=Bun.ms(R),P=Bun.ms(Q());console.log(q,x,z,A,B,C);console.log(E,F,G,J,K);console.log(L,P);function Q(){return"1m"}var R=1000;
"
`;
